### ISSUE TITLE
SGLD / SGHMC implementations

### ISSUE THREAD

**Issue by willtebbutt** [TuringLang/Turing.jl#516]
Both the SGLD and SGHMC implementations contains lines similar to the following in their `step` functions:
```julia
if verifygrad(grad)
    @debug "update latent variables..."
    v = zeros(Float64, size(old_θ))
    for k in 1:size(old_θ, 1)
        noise = rand(MvNormal(zeros(length(old_θ[k,:])), sqrt.(ϵ_t)*ones(length(old_θ[k,:]))))
        θ[k,:] = old_θ[k,:] - 0.5 * ϵ_t * grad[k,:] + noise
    end
end
```
It's not clear to me why we're indexing into `old_θ`, ` θ`, and `grad` using `[k, :]`. It's my understanding that all of these variables are `Vector`s, so their kth rows will always have length one. Is it possible for this not to be the case? Otherwise we should probably move to using `[k]` rather than `[k, :]` or broadcasting.

**Comment by yebai** [comment]
Hi @emilemathieu, will is refactoring your sghmc code. Can you help him out? Thanks!

**Comment by emilemathieu** [comment]
Hi !

`θ` are the parameters of the model - `v` and `grad` have the same dimension.
I believe `for k in 1:size(old_θ, 1)` loops though each parameters of the model, which can be a vector. I can't remind how the parameters are saved in memory though (concatenated vector? matrix?).
In any case you're right using `[k]` is better.

Cheers,
Emile

**Comment by willtebbutt** [comment]
Thanks @emilemathieu . This was my suspicion, I just wanted to ensure that there's nothing strange going on with the stochastic gradient stuff. Will close this when my current PR is merged.

**Comment by willtebbutt** [comment]
Closed by #515
